README.rst
Netresearch Session
===================

.. contents:: Inhaltsverzeichnis


Overview
========

Alternative session handler. Provides an interface to a more secure, faster and
scalable session data storage. And more secure session id handling.

Features
========

Uses caching framework to store session data.

You can configure every caching framework backend to store session data - a
preferred one would be redis or any other memory based caching system.

Validate generated session id against session storage for uniqueness.

Every newly generated session id is checked against session data storage for its
uniqueness to ensure security and data privacy.

Adds an XCLASS for tslib_feuserauth to overwrite session storage handling.

Configuration
=============

Session storage is configured like any other caching framework configuration.
Name of the used caching configuration is 'nr_session'.

You need to set up the nr_session storage configuration in your localconf.php:

(Example for use with Redis)

TYPO3 4.6 - 6.2
---------------

::

    $GLOBALS['$TYPO3_CONF_VARS']['SYS']['caching']['cacheConfigurations']
        ['nr_session'] = array(
            'backend'  => '\t3lib_cache_backend_RedisBackend',
            'frontend' => '\t3lib_cache_frontend_VariableFrontend',
            'options'  => array(
                'hostname'         => 'my.redis.host',
                //'port'             => 6379,
                'database'         => 0,
                //'password'         => '',
                //'compression'      => false,
                //'compressionLevel' => 1,
            ),
        );

TYPO3 4.5
---------

::

    $GLOBALS['$TYPO3_CONF_VARS']['SYS']['caching']['cacheConfigurations']
        ['nr_session'] = array(
            'backend'  => 't3lib_cache_backend_RedisBackend',
            'frontend' => 't3lib_cache_frontend_VariableFrontend',
            'options'  => array(
                'hostname'         => 'my.redis.host',
                //'port'             => 6379,
                'database'         => 0,
                //'password'         => '',
                //'compression'      => false,
                //'compressionLevel' => 1,
            ),
        );

You also need to set up your caching framework:

- http://wiki.typo3.org/Caching_framework#Enable_caching_framework_in_TYPO3_4.5_and_below


ChangeLog
2014-08-25  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Version: 2.0.1

2014-08-25  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Version: 2.0.0

2014-08-21  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Version: 1.0.0

2014-08-19  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* TYPO-2741: ensure newly generated session id is not in use already

	* TYPO-2741: outsource session handling into own extension nr_session



ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext "nr_session".
 *
 * Auto generated 25-08-2014 08:03
 *
 * Manual updates:
 * Only the data in the array - everything else is removed by next
 * writing. "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Netresearch session module.',
	'description' => 'Provides a more secure, stable and scalable session handler which uses caching framework.',
	'category' => 'fe',
	'author' => 'Sebastian Mendel',
	'author_company' => 'Netresearch GmbH & Co.KG',
	'author_email' => 'sebastian.mendel@netresearch.de',
	'constraints' => array(
		'depends' => array(
			'php' => '5.3.0-5.5.99',
			'typo3' => '4.5.0-6.2.99',
		),
		'conflicts' => array(
		),
		'suggests' => array(
			'nr_cache' => '1.0.4-',
			'nr_lock' => '1.0.0-',
		),
	),
	'state' => 'stable',
	'version' => '2.0.1',
	'_md5_values_when_last_written' => 'a:7:{s:9:"build.xml";s:4:"3f95";s:9:"ChangeLog";s:4:"c96d";s:16:"ext_autoload.php";s:4:"008b";s:12:"ext_icon.gif";s:4:"a459";s:17:"ext_localconf.php";s:4:"cba1";s:10:"README.rst";s:4:"b4fb";s:15:"src/Session.php";s:4:"c181";}',
);

?>

src/Session.php
<?php
declare(encoding = 'UTF-8');
/**
 * Holds session handler controller class.
 *
 * @category   Controller
 * @package    Netresearch
 * @subpackage Session
 * @author     Sebastian Mendel <sebastian.mendel@netresearch.de>
 * @license    AGPL http://www.netresearch.de/
 * @link       http://www.netresearch.de/
 */

namespace Netresearch\Session;

/**
 * Class Session.
 *
 * UX class for tslib_feUserAuth to write session data in caching framework.
 *
 * @category   Controller
 * @package    Netresearch
 * @subpackage Session
 * @author     Sebastian Mendel <sebastian.mendel@netresearch.de>
 * @license    AGPL http://www.netresearch.de/
 * @link       http://www.netresearch.de/
 */
class Session extends \tslib_feUserAuth
{

    /**
     * @var \t3lib_cache_frontend_VariableFrontend.
     */
    var $cache = null;

    /**
     * @var array holds session meta data.
     */
    var $arMeta = array();


    /**
     * Creates a user session record.
     *
     * @param array $arTempUser user data array
     *
     * @return void
     */
    function createUserSession($arTempUser)
    {
        /* @var \t3lib_db $TYPO3_DB */
        global $TYPO3_DB;

        \t3lib_div::devLog(
            'Create session ses_id = ' . $this->id, 'nr_session'
        );

        $this->arMeta = $this->getNewSessionRecord($arTempUser);
        $this->storeSessionData();

        // Updating lastLogin_column carrying information about last login.
        if ($this->lastLogin_column) {
            $TYPO3_DB->exec_UPDATEquery(
                $this->user_table,
                $this->userid_column . '=' . $TYPO3_DB->fullQuoteStr(
                    $arTempUser[$this->userid_column], $this->user_table
                ),
                array(
                    $this->lastLogin_column => $GLOBALS['EXEC_TIME'],
                )
            );
        }
    }



    /**
     * Read the user session from db.
     *
     * @param boolean $skipSessionUpdate Flag to skip session update.
     *
     * @return array user session data
     * @todo   implement IP lock; see parent::fetchUserSessionFromDB()
     * @todo   implement hash lock; see parent::fetchUserSessionFromDB()
     */
    function fetchUserSession($skipSessionUpdate = false)
    {
        \t3lib_div::devLog(
            'Fetch session ses_id = ' . $this->id, 'nr_session'
        );

        // load session
        $this->fetchSessionData();

        // load user with user_id from session
        $this->fetchUser($this->arMeta['ses_userid']);

        if (empty($this->arMeta)) {
            \t3lib_div::devLog(
                'log off, no session found', 'nr_session'
            );
            $this->logoff();
            return null;
        }

        if (empty($this->user)) {
            \t3lib_div::devLog(
                'log off, due to empty user', 'nr_session'
            );
            $this->logoff();
            return null;
        }

        $this->user += $this->arMeta;

        // A user was found
        if (intval($this->auth_timeout_field) > 0) {
            // Get timeout from object
            $timeout = intval($this->auth_timeout_field);
        } else {
            // Get timeout-time from user table
            $timeout = intval($this->user[$this->auth_timeout_field]);
        }

        // If timeout > 0 (TRUE) and current time has not exceeded the latest
        // sessions-time plus the timeout in seconds then accept user
        // Option later on: We could check that last update was at least x
        // seconds ago in order not to update twice in a row if one script
        // redirects to another...
        if ($timeout <= 0) {
            \t3lib_div::devLog(
                'log off, due to timeout of 0', 'nr_session'
            );
            $this->logoff();
            return null;
        }

        if ($GLOBALS['EXEC_TIME'] > ($this->user['ses_tstamp'] + $timeout)) {
            \t3lib_div::devLog(
                'log off, due to session timeout', 'nr_session'
            );
            $this->logoff();
            return null;
        }

        if (!$skipSessionUpdate) {
            $this->user[$this->auth_timeout_field]
                = $this->arMeta['ses_tstamp'] = $GLOBALS['EXEC_TIME'];
            $this->sesData_change = true;
            $this->storeSessionData();
        }

        return $this->user;
    }



    /**
     * Log out current user!
     * Removes the current session record, sets the internal ->user array to a
     * blank string; Thereby the current user (if any) is effectively logged out!
     *
     * We do not clear the session or remove its data.
     * This would break website functionality.
     * f. e. affiliate
     *
     * @return	void
     */
    function logoff()
    {
        \t3lib_div::devLog(
            'logoff: ' . $this->id, 'nr_session'
        );

        // Release the locked records
        \t3lib_BEfunc::lockRecords();

        $arHooks = &$GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['t3lib/class.t3lib_userauth.php'];

        // Hook for pre-processing the logoff() method, requested and implemented
        // by andreas.otto@dkd.de:
        if (is_array($arHooks['logoff_pre_processing'])) {
            $_params = array();
            foreach ($arHooks['logoff_pre_processing'] as $_funcRef) {
                if ($_funcRef) {
                    \t3lib_div::callUserFunction($_funcRef, $_params, $this);
                }
            }
        }

        $this->user = '';

        // Hook for post-processing the logoff() method, requested and implemented
        // by andreas.otto@dkd.de:
        if (is_array($arHooks['logoff_post_processing'])) {
            $_params = array();
            foreach ($arHooks['logoff_post_processing'] as $_funcRef) {
                if ($_funcRef) {
                    \t3lib_div::callUserFunction($_funcRef, $_params, $this);
                }
            }
        }
    }



    /**
     * The session_id is used to find user in the database.
     * Two tables are joined: The session-table with user_id of the session and
     * the usertable with its primary key
     * if the client is flash (e.g. from a flash application inside TYPO3 that
     * does a server request)
     * then don't evaluate with the hashLockClause, as the client/browser is
     * included in this hash
     * and thus, the flash request would be rejected
     *
     * @param integer $nUserId User UID
     *
     * @return void
     * @access private
     */
    protected function fetchUser($nUserId)
    {
        /* @var \t3lib_db $TYPO3_DB */
        global $TYPO3_DB;

        $this->user = $TYPO3_DB->exec_SELECTgetSingleRow(
            '*',
            $this->user_table,
            $this->userid_column . ' = ' . intval($nUserId)
            . ' ' . $this->user_where_clause()
        );

        if ($TYPO3_DB->sql_error()) {
            \t3lib_div::devLog(
                $TYPO3_DB->sql_error(),
                'nr_session',
                \t3lib_div::SYSLOG_SEVERITY_ERROR
            );
        }
    }



    /**
     * Sets the session data ($data) for $key and writes all session data
     * (from ->user['ses_data']) .
     * The data will last only for this login session since it is stored in the
     * session store.
     *
     * @param string $strKey Pointer to an associative key in the session data
     *                       array which is stored serialized in the field "ses_data"
     *                       of the session table.
     * @param mixed  $data   The variable to store in index $strKey
     *
     * @return void
     */
    function setAndSaveSessionData($strKey, $data)
    {
        $this->setKey('ses', $strKey, $data);
        $this->storeSessionData();
    }



    /**
     * Fetches the session data for the user (from configured caching backend)
     * based on the ->id of the current user-session.
     *
     * The session data is restored to $this->sesData.
     *
     * @return void
     * @see storeSessionData()
     */
    function fetchSessionData()
    {
        // Gets SesData if any AND if not already selected by session fixation
        // check in ->isExistingSessionRecord()
        if ($this->id && !count($this->sesData)) {

            \t3lib_div::devLog(
                'load session data', 'nr_session'
            );

            $sesDataRow = $this->cache()->get($this->id);

            if (is_array($sesDataRow)) {
                $this->arMeta               = $sesDataRow['meta'];
                $this->sesData              = $sesDataRow['content'];
                $this->sessionDataTimestamp = $sesDataRow['tstamp'];
            }
        }
    }



    /**
     * Will write UC and session data.
     * If the flag $this->userData_change has been set, the function ->writeUC
     * is called (which will save persistent user session data)
     * If the flag $this->sesData_change has been set, the fe_session_data table
     * is updated with the content of $this->sesData
     * If the $this->sessionDataTimestamp is NULL there was no session record yet,
     * so we need to insert it into the database
     *
     * @return void
     * @see fetchSessionData(), getKey(), setKey()
     */
    function storeSessionData()
    {
        // Saves UC and SesData if changed.
        if ($this->userData_change) {
            $this->writeUC('');
            $this->userData_change = false;
        }

        if (empty($this->sesData_change)) {
            // no change in session data
            return;
        }

        if (empty($this->id)) {
            // no session id
            return;
        }

        $this->sessionDataTimestamp = $GLOBALS['EXEC_TIME'];

        \t3lib_div::devLog(
            'Set session data: "' . $this->id . '"',
            'nr_session'
        );
        $this->cache()->set(
            $this->id,
            array(
                'meta'    => $this->arMeta,
                'content' => $this->sesData,
                'tstamp'  => $GLOBALS['EXEC_TIME'],
            )
        );

        $this->setSessionCookie();

        $this->sesData_change = false;
    }



    /**
     * Removes data of the current session.
     *
     * @return void
     */
    public function removeSessionData()
    {
        \t3lib_div::devLog(
            'Drop session data: "' . $this->id . '"',
            'nr_session'
        );
        $this->cache()->remove($this->id);
    }



    /**
     * Executes the garbage collection of session data and session.
     * The lifetime of session data is defined by
     * $TYPO3_CONF_VARS['FE']['sessionDataLifetime'].
     *
     * @return    void
     */
    public function gc()
    {
    }



    /**
     * Determine whether there's an according session record to a given session_id.
     * Don't care if session record is still valid or not.
     *
     * @param string $strId Claimed Session ID
     *
     * @return boolean Returns TRUE if a corresponding session was found
     */
    function isExistingSessionRecord($strId)
    {
        return $this->cache()->has($strId);
    }



    /**
     * Returns cache frontend controller.
     *
     * @return \t3lib_cache_frontend_VariableFrontend
     */
    private function cache()
    {
        if (null !== $this->cache) {
            return $this->cache;
        }

        /** @var \t3lib_cache_Manager $typo3CacheManager */
        global $typo3CacheManager;

        \t3lib_cache::initializeCachingFramework();

        try {
            $this->cache = $typo3CacheManager->getCache(
                'nr_session'
            );
        } catch (\t3lib_cache_exception_NoSuchCache $e) {
            /** @var array[] $TYPO3_CONF_VARS */
            global $TYPO3_CONF_VARS;
            $arCacheConfigurations
                = $TYPO3_CONF_VARS['SYS']['caching']['cacheConfigurations'];

            /** @var \t3lib_cache_Factory $typo3CacheFactory */
            global $typo3CacheFactory;
            $this->cache = $typo3CacheFactory->create(
                'nr_session',
                $arCacheConfigurations['nr_session']['frontend'],
                $arCacheConfigurations['nr_session']['backend'],
                $arCacheConfigurations['nr_session']['options']
            );
        }

        return $this->cache;
    }



    /**
     * Creates a new session ID.
     *
     * @return	string		The new session ID
     */
    public function createSessionId()
    {
        $isSessionIdClean = false;
        $nRetries         = 0;

        do {
            if ($nRetries > 1) {
                // we retry only 1 time - throwing the exception gives the system
                // some time to get more healthy and generate new clean session ids
                \t3lib_div::devLog(
                    'Could not get secure and empty session id.',
                    'nr_session',
                    \t3lib_div::SYSLOG_SEVERITY_ERROR
                );
                throw new \RuntimeException(
                    'Could not get secure and empty session id.'
                );
            }

            $strSessionId = parent::createSessionId();

            /** @var \t3lib_lock $lock */
            $lock = \t3lib_div::makeInstance(
                't3lib_lock', $strSessionId, 'simple'
            );

            if ($lock->acquire()) {
                // acquire returns true for immediately retrieved locks
                // means no other process held the lock before
                if ($this->isExistingSessionRecord($strSessionId)) {
                    // there is already session data associated with this session id
                    // so we mark this id as dirty
                    $isSessionIdClean = false;
                } else {
                    // there is no other data associated with this session id
                    $isSessionIdClean = true;

                    // create a cache entry for this session
                    $this->cache()->set(
                        $strSessionId,
                        array(
                            'meta'    => $this->arMeta,
                            'content' => $this->sesData,
                            'tstamp'  => $GLOBALS['EXEC_TIME'],
                        )
                    );
                }
            } else {
                // could not get lock, or not immediately - so it is unlikely
                // this session data is clean
                $isSessionIdClean = false;
            }
            $lock->release();
            $nRetries++;
        } while (false === $isSessionIdClean);

        return $strSessionId;
    }
}

?>


ext_localconf.php
<?php
/**
 * Extension config script
 *
 * PHP version 5
 *
 * @category   Netresearch
 * @package    Session
 * @subpackage Configuration
 * @author     Sebastian Mendel <sebastian.mendel@netresearch.de>
 * @license    AGPL http://www.netresearch.de/
 * @link       http://www.netresearch.de/
 */

defined('TYPO3_MODE') or die('Access denied.');

global $TYPO3_CONF_VARS;

$arCacheCfg = &$TYPO3_CONF_VARS['SYS']['caching']['cacheConfigurations'];

if (! empty($arCacheCfg['nr_session'])) {
    if (empty($arCacheCfg['nr_session']['frontend'])) {
        $arCacheCfg['nr_session']['frontend']
            = '\t3lib_cache_frontend_VariableFrontend';
    }

    $arCacheCfg['nr_session']['options']['defaultLifetime']
        = $TYPO3_CONF_VARS['FE']['sessionDataLifetime'];


    // TYPO3 4.5 - 4.7
    // register XCLASS to overwrite session storage handling
    $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['tslib/class.tslib_feuserauth.php']
        = t3lib_extMgm::extPath('nr_session') . 'src/Session.php';

    require_once t3lib_extMgm::extPath('nr_session') . 'src/Session.php';
    class ux_tslib_feUserAuth extends \Netresearch\Session\Session {}


    // TYPO3 6.2
    $GLOBALS['TYPO3_CONF_VARS']['SYS']['Objects']
        ['TYPO3\CMS\Frontend\Authentication\FrontendUserAuthentication']
        ['className'] = 'Netresearch\Session\Session';
}


?>
